pkg/util/pki/renewaltime.go

```diff
@@ -57,3 +52,43 @@ func RenewalTime(notBefore, notAfter time.Time, renewBeforeOverride *metav1.Dura
 	rt := metav1.NewTime(notAfter.Add(-1 * renewBefore).Truncate(time.Second))
 	return &rt
 }
+
+func RenewBefore(actualDuration time.Duration, renewBeforeOverride *string) time.Duration {
+	defaultRenewBefore := actualDuration / 3
+
+	if renewBeforeOverride == nil {
+		return defaultRenewBefore
+	}
+
+	d, pct, err := ParseDurationOrPercent(*renewBeforeOverride)
+	if err != nil {
+		return defaultRenewBefore
+	}
+
+	// If spec.renewBefore was set (and is less than duration)
+	// respect that. We don't want to prevent users from renewing
+	// longer lived certs more frequently.
+	if d != nil && *d > 0 && *d < actualDuration {
+		return *d
+	} else if pct != nil && *pct > 0 && *pct < 100 {
+		return actualDuration * time.Duration(100-*pct) / 100
+	}
+
+	return defaultRenewBefore
+}
+
+func ParseDurationOrPercent(s string) (*time.Duration, *int, error) {
```

Comment:
Could you add some tests to this function to make sure the error messages are understandable?